docs/proposed/psa-driver-interface.md

```diff
@@ -321,6 +321,148 @@ TODO: key input and output for opaque drivers; deterministic key generation for
 
 TODO
 
+### Driver entry points for PAKE
+
+PAKE operation is divided into two stages: collecting inputs and computation. Core side is responsible for keeping inputs and core set-data functions do not have driver entry points. Collected inputs are available for drivers via get-data functions for `password`, `role` and `cipher_suite`.
+
+### PAKE driver dispatch logic
+The core decides whether to dispatch a PAKE operation to a driver based on the location of the provided password.
+When all inputs are collected and `"psa_pake_output"` or `"psa_pake_input"` is called for the first time `"pake_setup"` driver entry point is invoked.
+
+1. Lifetime of the `password` is local storage
+- if there is a transparent driver available for the given configuration, the core calls that driver's `"pake_setup"` and subsequent entry points.
+- if a transparent driver is not available or can not handle a given configuration, the core uses its built-in implementation.
+2. Lifetime of the `password` is test driver
+- the core calls opaque driver's `"pake_setup"` and subsequent entry points.
+
+### Summary of entry points for PAKE
+
+A PAKE driver has the following entry points:
+* `"pake_setup"` (mandatory): always the first entry point to be called. It is called when all inputs are collected and the computation stage starts. 
+* `"pake_output"` (mandatory): derive cryptographic material for the specified step and output it.
+* `"pake_input"` (mandatory): provides cryptographic material in the format appropriate for the specified step.
+* `"pake_get_implicit_key"` (mandatory): returns implicitly confirmed shared secret from a PAKE.
+* `"pake_abort"` (mandatory): always the last entry point to be called.
+
+For naming purposes, here and in the following subsection, this specification takes the example of a driver with the prefix `"acme"` that implements the PAKE entry point family with a capability that does not use the `"names"` property to declare different type and entry point names. Such a driver must implement the following type and functions, as well as the entry points listed above and described in the following subsections:
+```
+typedef ... acme_pake_operation_t;
+psa_status_t acme_pake_abort( acme_pake_operation_t *operation );
+```
+
+#### PAKE driver inputs
+
+The core conveys the initial inputs for a PAKE operation via an opaque data structure of type `psa_crypto_driver_pake_inputs_t`.
+
+```
+typedef ... psa_crypto_driver_pake_inputs_t; // implementation-specific type
+```
+
+A driver receiving an argument that points to a `psa_crypto_driver_pake_inputs_t` can retrieve its contents by calling one of the get-data functions below.
+
+```
+psa_status_t psa_crypto_driver_pake_get_password_len(
```

Comment:
I don't know if it's spelled out explicitly anywhere, but I'm assuming that it's allowed for an opaque driver to support multiple key locations (for example one location for volatile keys inside of the SE, and one location for wrapped keys). In that case, the driver needs to be able to figure out the location to use for the password. For other driver entry points, this is enabled by the key attributes passed alongside the key data. I'm thinking that we need to add something here such that the driver can figure out what the location of the password.

Reply:
Currently core side takes only one password per operation and driver just gets key data via getter function. Probably I'm missing something, but as far as I understand for other key we need to use new operation.
If key location/lifetime of the password is really needed I can add getter function for this as core already have this information in `inputs`.

Reply:
The same driver (as in, the same code) can support multiple _[locations](https://arm-software.github.io/psa-api/crypto/1.1/api/keys/lifetimes.html#c.psa_key_location_t)_, e.g. a phone with dual SIM with the driver supporting both SIM. A driver responsible for a given location is systematically responsible for multiple lifetimes, namely, all the [persistence levels](https://arm-software.github.io/psa-api/crypto/1.1/api/keys/lifetimes.html#c.psa_key_persistence_t) (e.g. volatile, stored, factory-set-and-immutable).

The [generic way driver entry points for operations are formed](https://github.com/Mbed-TLS/mbedtls/blob/development/docs/proposed/psa-driver-interface.md#overview-of-driver-entry-points) is by taking API functions and replacing `psa_key_id_t` by `attributes, key_buffer, key_buffer_size`. So indeed `psa_crypto_driver_pake_get_password` should have an `attributes` parameter.

I made the same mistake as @mprse with `psa_crypto_driver_key_derivation_get_input_key` — [noted](https://github.com/Mbed-TLS/mbedtls/pull/5451/files#r1082947887).

Reply:
Thank you for the details. Looking at location and lifetime there are the following main options:
Location: local storage/secure element
Lifetime: volatile/persistent

Now for example we have the following main combinations:
- (local storage, volatile) indicates that the key is stored in our device volatile memory like RAM.
- (local storage, persistent) indicates that the key is stored in our device non-volatile memory like Flash.
- (secure element, volatile) indicates that the key is stored in external device volatile memory (not sure if this is possible).
- (secure element, persistent) indicates that the key is stored in external device non-volatile memory.

We have separate drivers for different locations, but now information about key lifetime is missing. What I don't understand is why driver needs this information. It gets key data from core anyway and just need to use it.

Reply:
Secure element, volatile is possible. Not very common, but we have had requests for it. Mainly, secure elements are used to provide a higher level of security for long-term keys, and long-term keys are persistent. But using a volatile key with a secure element can be useful sometimes. In particular, it allows using the secure element as an accelerator, so the non-SE part doesn't need to support the algorithm at all. You could do the same with an accelerator driver, but if you already have the secure element driver and you're ok with tweaking your application code, you save code size by not defining an accelerator driver.

The driver needs to know the full lifetime information: the location in case the driver supports multiple location, and the persistence level so that it knows where to load/store the key data (RAM, or potentially multiple storage partitions with different persistence levels).